aten/src/ATen/native/xnnpack/MaxPooling.cpp

```diff
@@ -0,0 +1,212 @@
+  // * Kernel must be a 2D IntArrayRef containing two positive numbers.
+  //   Furthermore, 1x1 kernels are neither valid nor meaningful.
+  // * Padding must be a 2D IntArrayRef containing two non-negative numbers.
+  // * Stride must be a 2D IntArrayRef containing two positive numbers.
+  // * Dilation must be a 2D IntArrayRef containing two positive numbers.
+  // * Ceil mode is not supported and must be disabled.
+  // * output_max must be greater than output_min.
+  //   Namely, setting both output_min and output_max to 0 is not valid usage.
+  // * Finally, application of this operator to the input tensor with the given
+  //   max pool 2d parameters must result in an output tensor with a valid shape.
+
+  return xnnpack::internal::available() &&
+      // Input
+      (4 == input.dim()) &&
+      (c10::DeviceType::CPU == input.device().type()) &&
+      (kFloat == input.scalar_type()) &&
+      !input.requires_grad() &&
+      // Kernel
+      (2 == parameters.kernel.size()) &&
+      (parameters.kernel[Layout::Parameter::height] > 0) &&
+      (parameters.kernel[Layout::Parameter::width] > 0) &&
+      ((parameters.kernel[Layout::Parameter::height] *
+        parameters.kernel[Layout::Parameter::width]) > 1) &&
+      // Padding
+      (2 == parameters.padding.size()) &&
+      (parameters.padding[Layout::Parameter::height] >= 0) &&
+      (parameters.padding[Layout::Parameter::width] >= 0) &&
+      // Stride
+      (2 == parameters.stride.size()) &&
+      (parameters.stride[Layout::Parameter::height] > 0) &&
+      (parameters.stride[Layout::Parameter::width] > 0) &&
+      // Dilation
+      (2 == parameters.dilation.size()) &&
+      (parameters.dilation[Layout::Parameter::height] > 0) &&
+      (parameters.dilation[Layout::Parameter::width] > 0) &&
+      // Ceil Mode
+      !ceil_mode &&
+      // Output Min / Max
+      (output_max > output_min) &&
+      // Output
+      (pooling_output_shape(
+        input.size(Layout::Activation4D::height),
+        parameters.kernel[Layout::Parameter::height],
+        parameters.padding[Layout::Parameter::height],
+        parameters.stride[Layout::Parameter::height],
+        parameters.dilation[Layout::Parameter::height],
+        ceil_mode) > 0) &&
+      (pooling_output_shape(
+        input.size(Layout::Activation4D::width),
+        parameters.kernel[Layout::Parameter::width],
+        parameters.padding[Layout::Parameter::width],
+        parameters.stride[Layout::Parameter::width],
+        parameters.dilation[Layout::Parameter::width],
+        ceil_mode) > 0) &&
+      true;
+}
+
+Tensor max_pool2d(
+    const Tensor& input,
+    const IntArrayRef kernel_,
+    const IntArrayRef padding_,
+    IntArrayRef stride_,
+    const IntArrayRef dilation_,
+    const bool ceil_mode,
+    const float output_min,
+    const float output_max) {
+  using namespace internal;
+
+  // A call to max_pool2d must have been gated by a call to use_maxpool2d, so
```

Comment:
Can this be enforced somehow? That is, if you come here without going through use_maxpool2d, then it is caught? And if it is not possible then it would be good to couple these two together such that this is not possible. I guess the trouble would be the API of max_pool2d. Maybe it can return optional Tensor?

Reply:
I cannot think of a way since both these functions are pure and stateless.

Reply:
I think the idea is that you would remove `use_max_pool2d` as a separate function and bring all that logic into `max_pool2d`.  That function would first perform the checks.  If they fail, it would return `nullopt`, and if they succeed, it would perform the max pool and return the result tensor.  Then the calling code would change from

```
if (can_use_xnnpack()) {
  return use_xnnpack();
}
```
to
```
optional<Tensor> ret = use_xnnpack();
if (ret) {
  return *ret;
}
```

It's not clear to me whether this would be better or worse than the current implementation.